README.md
# Retro LED display
This project started in the mids of the 80th when a good friend of mine approached me, asking if we could build an LED display for his birthday party. So we did. Nobody believed us that we did it ourselves.

![LED array display](display.png)

Technically, it was mounted on an old wooden panel, and  driven by a self-designed and build hardware extension for by my C64. the software written in 6502 (actually 6510) assembler, using the character set of the C64. It implemented a bunch of special effects as well, like blending randomly one text to another, scrolling in different directions etc.
Later, the hardware was reused for other projects but the display itself with its about 500 LEDs. The project died with my C64.

A couple of years ago I found it it back from my parents house, and recovered it. Only recently I made a first attempt to develop some new hardware and interface it with a RaspberryPi.

Maybe somebody might find it interesting. Here's an example of how it looks like:

![Hello You example](hello_you.png)

## Hardware
The display consist of 504 red LEDs (state of the art mid 80th). The exact specifications of the LEDs got lost. It's organised in 3 blocks of 7 rows and 24 (or 3x8) columns each. More details can be found in the hardware folder.

## Software
The driver software is written in C using libgpiod. Check the README.md file in the software folder.


software/ChangeLog
20200823 initial port to libgpiod



software/src/common.c
/**********************************************************************
* Filename    : common.c
* Description : Common methods
* License     : GPL2
* Author      : Ulrich Schwickerath, 2020
**********************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <string.h>
#include <unistd.h>
#include "common.h"

unsigned char mirror_byte(unsigned char byte){
  /* returns a mirrored version of byte */
  byte = (byte & 0xF0) >> 4 | (byte & 0x0F) << 4;
  byte = (byte & 0xCC) >> 2 | (byte & 0x33) << 2;
  byte = (byte & 0xAA) >> 1 | (byte & 0x55) << 1;
  return byte;
}

void create_display_thread(void *display_function) {
  /* create the display thread */
  pthread_t displaythread;
  int rc = pthread_create(&displaythread, NULL, display_function, (void *) &display_buffer[0]);
  if (rc) {
    printf("Cannot create the display thread\n");
    exit(-1);
  }
  sleep(1);
}

void write_ascii(unsigned char *buf, unsigned char *source, unsigned int asciicode){
  /* render a mirrored character based on its ASCII code */
  for (int i=0; i<8; i++){
    unsigned char byte = mirror_byte(*(source + 8 * asciicode + i));
    *(buf+i) = byte;
  }
}

void write_string(int columns, unsigned char *buf, unsigned char *charset, char *text){
  /* write a string into a buffer 
  Note:  int offset = (int)'@'; */
  int offset = 0;
  int flag = 0;
  for (int i = 0; i < columns; i++) {
    int num = (int) *(text + i);
    if (num == 0 || flag) {
      num = 32;
      flag = 1;
    } else {
      num = num - offset;
    }
    write_ascii(buf + ROWS * i, charset, num);
  }
}

void shiftleftonce(int row, int rows, int columns, unsigned char *buf){
  /* scroll the text left by one bit */
  int i, j;
  int ov[columns];
  
  for (j = 0; j < columns; j++) {
    /* create the overflow */
    i = row + j * rows;
    if ((buf[i] & 0x80) == 0) {
      ov[j] = 0;
    } else {
      ov[j] = 1;
    }
  }
  for (j = 0; j < columns; j++) {
    i = row + j * rows;
    if (j < columns - 1) {
      buf[i] = (buf[i] << 1) | ov[j + 1];
    } else {
      buf[i] = buf[i] << 1;
    }
  }
}


software/src/simulate.c
/**********************************************************************
* Filename    : simulate.c
* Description : simulate the output on the terminal
* License     : GPL2
* Author      : Ulrich Schwickerath, 2020
**********************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <string.h>
#include <unistd.h>
#include "common.h"

extern unsigned char* font[];

void show_char(unsigned char x){
  int i;
  unsigned char mask = 0x80;
  for (i=0; i<8; i++){
    if ( x & mask >> i ){
      printf("*");
    } else {
      printf(" ");
    }
  }
}

void * thread_show_buffer_on_terminal(void * buff){
  int i, j;
  int nx = ROWS;
  int ny = COLUMNS;
  while (1){
    printf("\e[1;1H\e[2J");
    for (j=0; j<nx; j++){
      for (i=0; i<ny; i++){
	unsigned char byte = *(((unsigned char *)buff)+ROWS*i + j);
	show_char(byte);
      }
      printf("\n");
    }
    usleep(100000);
  }
}


int main(void) {
  
  unsigned char characterset[CHARSET_SIZE+1];
  char textbuffer[LCOLUMNS];
  int ret;
  // initialise text
  memcpy(characterset, font, 1024);

  // create the display thread
  create_display_thread(thread_show_buffer_on_terminal);

  // write some text into the large buffer
  strcpy(textbuffer, "Das Pferd frisst keinen Gurkensalat        ");
  write_string(strlen(textbuffer), (unsigned char*)large_display_buffer, characterset, textbuffer);
  // make a sliding window
  int copy = COLUMNS*ROWS; // fill the full display buffer
  ret = system("/usr/bin/clear");
  if (ret == 0){
    memcpy((unsigned char*)display_buffer, ((unsigned char*)large_display_buffer), copy);
    for (int i=0;i<strlen(textbuffer)*ROWS;i++){
      for (int ii=0;ii<8;ii++){
	for (int jj=0;jj<ROWS;jj++){
	  shiftleftonce(jj, ROWS, LCOLUMNS, ((unsigned char*)large_display_buffer));
	}
	memcpy((unsigned char*)display_buffer, ((unsigned char*)large_display_buffer), copy);
	usleep(50000);
      }
    }
  }
}


software/include/font8x8_basic.h
/**
 * Taken from https://github.com/dhepper/font8x8/blob/master/font8x8_basic.h
 *
 * 8x8 monochrome bitmap fonts for rendering
 * Author: Daniel Hepper <daniel@hepper.net>
 *
 * License: Public Domain
 *
 * Based on:
 * // Summary: font8x8.h
 * // 8x8 monochrome bitmap fonts for rendering
 * //
 * // Author:
 * //     Marcel Sondaar
 * //     International Business Machines (public domain VGA fonts)
 * //
 * // License:
 * //     Public Domain
 *
 * Fetched from: http://dimensionalrift.homelinux.net/combuster/mos3/?p=viewsource&file=/modules/gfx/font8_8.asm
 **/

// Constant: font8x8_basic
// Contains an 8x8 font map for unicode points U+0000 - U+007F (basic latin)
char font[128][8] = {
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0000 (nul)
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0001
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0002
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0003
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0004
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0005
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0006
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0007
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0008
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0009
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+000A
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+000B
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+000C
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+000D
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+000E
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+000F
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0010
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0011
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0012
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0013
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0014
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0015
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0016
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0017
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0018
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0019
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+001A
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+001B
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+001C
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+001D
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+001E
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+001F
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0020 (space)
    { 0x18, 0x3C, 0x3C, 0x18, 0x18, 0x00, 0x18, 0x00},   // U+0021 (!)
    { 0x36, 0x36, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0022 (")
    { 0x36, 0x36, 0x7F, 0x36, 0x7F, 0x36, 0x36, 0x00},   // U+0023 (#)
    { 0x0C, 0x3E, 0x03, 0x1E, 0x30, 0x1F, 0x0C, 0x00},   // U+0024 ($)
    { 0x00, 0x63, 0x33, 0x18, 0x0C, 0x66, 0x63, 0x00},   // U+0025 (%)
    { 0x1C, 0x36, 0x1C, 0x6E, 0x3B, 0x33, 0x6E, 0x00},   // U+0026 (&)
    { 0x06, 0x06, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0027 (')
    { 0x18, 0x0C, 0x06, 0x06, 0x06, 0x0C, 0x18, 0x00},   // U+0028 (()
    { 0x06, 0x0C, 0x18, 0x18, 0x18, 0x0C, 0x06, 0x00},   // U+0029 ())
    { 0x00, 0x66, 0x3C, 0xFF, 0x3C, 0x66, 0x00, 0x00},   // U+002A (*)
    { 0x00, 0x0C, 0x0C, 0x3F, 0x0C, 0x0C, 0x00, 0x00},   // U+002B (+)
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x0C, 0x0C, 0x06},   // U+002C (,)
    { 0x00, 0x00, 0x00, 0x3F, 0x00, 0x00, 0x00, 0x00},   // U+002D (-)
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x0C, 0x0C, 0x00},   // U+002E (.)
    { 0x60, 0x30, 0x18, 0x0C, 0x06, 0x03, 0x01, 0x00},   // U+002F (/)
    { 0x3E, 0x63, 0x73, 0x7B, 0x6F, 0x67, 0x3E, 0x00},   // U+0030 (0)
    { 0x0C, 0x0E, 0x0C, 0x0C, 0x0C, 0x0C, 0x3F, 0x00},   // U+0031 (1)
    { 0x1E, 0x33, 0x30, 0x1C, 0x06, 0x33, 0x3F, 0x00},   // U+0032 (2)
    { 0x1E, 0x33, 0x30, 0x1C, 0x30, 0x33, 0x1E, 0x00},   // U+0033 (3)
    { 0x38, 0x3C, 0x36, 0x33, 0x7F, 0x30, 0x78, 0x00},   // U+0034 (4)
    { 0x3F, 0x03, 0x1F, 0x30, 0x30, 0x33, 0x1E, 0x00},   // U+0035 (5)
    { 0x1C, 0x06, 0x03, 0x1F, 0x33, 0x33, 0x1E, 0x00},   // U+0036 (6)
    { 0x3F, 0x33, 0x30, 0x18, 0x0C, 0x0C, 0x0C, 0x00},   // U+0037 (7)
    { 0x1E, 0x33, 0x33, 0x1E, 0x33, 0x33, 0x1E, 0x00},   // U+0038 (8)
    { 0x1E, 0x33, 0x33, 0x3E, 0x30, 0x18, 0x0E, 0x00},   // U+0039 (9)
    { 0x00, 0x0C, 0x0C, 0x00, 0x00, 0x0C, 0x0C, 0x00},   // U+003A (:)
    { 0x00, 0x0C, 0x0C, 0x00, 0x00, 0x0C, 0x0C, 0x06},   // U+003B (;)
    { 0x18, 0x0C, 0x06, 0x03, 0x06, 0x0C, 0x18, 0x00},   // U+003C (<)
    { 0x00, 0x00, 0x3F, 0x00, 0x00, 0x3F, 0x00, 0x00},   // U+003D (=)
    { 0x06, 0x0C, 0x18, 0x30, 0x18, 0x0C, 0x06, 0x00},   // U+003E (>)
    { 0x1E, 0x33, 0x30, 0x18, 0x0C, 0x00, 0x0C, 0x00},   // U+003F (?)
    { 0x3E, 0x63, 0x7B, 0x7B, 0x7B, 0x03, 0x1E, 0x00},   // U+0040 (@)
    { 0x0C, 0x1E, 0x33, 0x33, 0x3F, 0x33, 0x33, 0x00},   // U+0041 (A)
    { 0x3F, 0x66, 0x66, 0x3E, 0x66, 0x66, 0x3F, 0x00},   // U+0042 (B)
    { 0x3C, 0x66, 0x03, 0x03, 0x03, 0x66, 0x3C, 0x00},   // U+0043 (C)
    { 0x1F, 0x36, 0x66, 0x66, 0x66, 0x36, 0x1F, 0x00},   // U+0044 (D)
    { 0x7F, 0x46, 0x16, 0x1E, 0x16, 0x46, 0x7F, 0x00},   // U+0045 (E)
    { 0x7F, 0x46, 0x16, 0x1E, 0x16, 0x06, 0x0F, 0x00},   // U+0046 (F)
    { 0x3C, 0x66, 0x03, 0x03, 0x73, 0x66, 0x7C, 0x00},   // U+0047 (G)
    { 0x33, 0x33, 0x33, 0x3F, 0x33, 0x33, 0x33, 0x00},   // U+0048 (H)
    { 0x1E, 0x0C, 0x0C, 0x0C, 0x0C, 0x0C, 0x1E, 0x00},   // U+0049 (I)
    { 0x78, 0x30, 0x30, 0x30, 0x33, 0x33, 0x1E, 0x00},   // U+004A (J)
    { 0x67, 0x66, 0x36, 0x1E, 0x36, 0x66, 0x67, 0x00},   // U+004B (K)
    { 0x0F, 0x06, 0x06, 0x06, 0x46, 0x66, 0x7F, 0x00},   // U+004C (L)
    { 0x63, 0x77, 0x7F, 0x7F, 0x6B, 0x63, 0x63, 0x00},   // U+004D (M)
    { 0x63, 0x67, 0x6F, 0x7B, 0x73, 0x63, 0x63, 0x00},   // U+004E (N)
    { 0x1C, 0x36, 0x63, 0x63, 0x63, 0x36, 0x1C, 0x00},   // U+004F (O)
    { 0x3F, 0x66, 0x66, 0x3E, 0x06, 0x06, 0x0F, 0x00},   // U+0050 (P)
    { 0x1E, 0x33, 0x33, 0x33, 0x3B, 0x1E, 0x38, 0x00},   // U+0051 (Q)
    { 0x3F, 0x66, 0x66, 0x3E, 0x36, 0x66, 0x67, 0x00},   // U+0052 (R)
    { 0x1E, 0x33, 0x07, 0x0E, 0x38, 0x33, 0x1E, 0x00},   // U+0053 (S)
    { 0x3F, 0x2D, 0x0C, 0x0C, 0x0C, 0x0C, 0x1E, 0x00},   // U+0054 (T)
    { 0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x3F, 0x00},   // U+0055 (U)
    { 0x33, 0x33, 0x33, 0x33, 0x33, 0x1E, 0x0C, 0x00},   // U+0056 (V)
    { 0x63, 0x63, 0x63, 0x6B, 0x7F, 0x77, 0x63, 0x00},   // U+0057 (W)
    { 0x63, 0x63, 0x36, 0x1C, 0x1C, 0x36, 0x63, 0x00},   // U+0058 (X)
    { 0x33, 0x33, 0x33, 0x1E, 0x0C, 0x0C, 0x1E, 0x00},   // U+0059 (Y)
    { 0x7F, 0x63, 0x31, 0x18, 0x4C, 0x66, 0x7F, 0x00},   // U+005A (Z)
    { 0x1E, 0x06, 0x06, 0x06, 0x06, 0x06, 0x1E, 0x00},   // U+005B ([)
    { 0x03, 0x06, 0x0C, 0x18, 0x30, 0x60, 0x40, 0x00},   // U+005C (\)
    { 0x1E, 0x18, 0x18, 0x18, 0x18, 0x18, 0x1E, 0x00},   // U+005D (])
    { 0x08, 0x1C, 0x36, 0x63, 0x00, 0x00, 0x00, 0x00},   // U+005E (^)
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF},   // U+005F (_)
    { 0x0C, 0x0C, 0x18, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+0060 (`)
    { 0x00, 0x00, 0x1E, 0x30, 0x3E, 0x33, 0x6E, 0x00},   // U+0061 (a)
    { 0x07, 0x06, 0x06, 0x3E, 0x66, 0x66, 0x3B, 0x00},   // U+0062 (b)
    { 0x00, 0x00, 0x1E, 0x33, 0x03, 0x33, 0x1E, 0x00},   // U+0063 (c)
    { 0x38, 0x30, 0x30, 0x3e, 0x33, 0x33, 0x6E, 0x00},   // U+0064 (d)
    { 0x00, 0x00, 0x1E, 0x33, 0x3f, 0x03, 0x1E, 0x00},   // U+0065 (e)
    { 0x1C, 0x36, 0x06, 0x0f, 0x06, 0x06, 0x0F, 0x00},   // U+0066 (f)
    { 0x00, 0x00, 0x6E, 0x33, 0x33, 0x3E, 0x30, 0x1F},   // U+0067 (g)
    { 0x07, 0x06, 0x36, 0x6E, 0x66, 0x66, 0x67, 0x00},   // U+0068 (h)
    { 0x0C, 0x00, 0x0E, 0x0C, 0x0C, 0x0C, 0x1E, 0x00},   // U+0069 (i)
    { 0x30, 0x00, 0x30, 0x30, 0x30, 0x33, 0x33, 0x1E},   // U+006A (j)
    { 0x07, 0x06, 0x66, 0x36, 0x1E, 0x36, 0x67, 0x00},   // U+006B (k)
    { 0x0E, 0x0C, 0x0C, 0x0C, 0x0C, 0x0C, 0x1E, 0x00},   // U+006C (l)
    { 0x00, 0x00, 0x33, 0x7F, 0x7F, 0x6B, 0x63, 0x00},   // U+006D (m)
    { 0x00, 0x00, 0x1F, 0x33, 0x33, 0x33, 0x33, 0x00},   // U+006E (n)
    { 0x00, 0x00, 0x1E, 0x33, 0x33, 0x33, 0x1E, 0x00},   // U+006F (o)
    { 0x00, 0x00, 0x3B, 0x66, 0x66, 0x3E, 0x06, 0x0F},   // U+0070 (p)
    { 0x00, 0x00, 0x6E, 0x33, 0x33, 0x3E, 0x30, 0x78},   // U+0071 (q)
    { 0x00, 0x00, 0x3B, 0x6E, 0x66, 0x06, 0x0F, 0x00},   // U+0072 (r)
    { 0x00, 0x00, 0x3E, 0x03, 0x1E, 0x30, 0x1F, 0x00},   // U+0073 (s)
    { 0x08, 0x0C, 0x3E, 0x0C, 0x0C, 0x2C, 0x18, 0x00},   // U+0074 (t)
    { 0x00, 0x00, 0x33, 0x33, 0x33, 0x33, 0x6E, 0x00},   // U+0075 (u)
    { 0x00, 0x00, 0x33, 0x33, 0x33, 0x1E, 0x0C, 0x00},   // U+0076 (v)
    { 0x00, 0x00, 0x63, 0x6B, 0x7F, 0x7F, 0x36, 0x00},   // U+0077 (w)
    { 0x00, 0x00, 0x63, 0x36, 0x1C, 0x36, 0x63, 0x00},   // U+0078 (x)
    { 0x00, 0x00, 0x33, 0x33, 0x33, 0x3E, 0x30, 0x1F},   // U+0079 (y)
    { 0x00, 0x00, 0x3F, 0x19, 0x0C, 0x26, 0x3F, 0x00},   // U+007A (z)
    { 0x38, 0x0C, 0x0C, 0x07, 0x0C, 0x0C, 0x38, 0x00},   // U+007B ({)
    { 0x18, 0x18, 0x18, 0x00, 0x18, 0x18, 0x18, 0x00},   // U+007C (|)
    { 0x07, 0x0C, 0x0C, 0x38, 0x0C, 0x0C, 0x07, 0x00},   // U+007D (})
    { 0x6E, 0x3B, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},   // U+007E (~)
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}    // U+007F
};


software/src/ledsoftware.c
/**********************************************************************
* Filename    : LEDsoftware.c
* Description : Control LED by 74HC595
* License     : GPL2
* Author      : Ulrich Schwickerath, 2020
**********************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <gpiod.h>
#include "common.h"

#ifndef CONSUMER
#define CONSUMER "LEDband"
#endif

#define dataPin 1        // DS pin of 74HC595(Pin14)
#define columnclockPin 5 // CH_CP pin of 74HC595(Pin11)
#define latchPin 26      // ST_CP pin of 74HC595(Pin12)
#define clrbarPin 4      // clear pin
#define fePin 6          // FE pin
#define rowdataPin 27    // block select pin
#define rowclockPin 28   // block clock pin


extern unsigned char *font[];

unsigned int LOW = 0;
unsigned int HIGH = 0xffffff;

struct gpiod_chip *chip;
struct gpiod_line *data;
struct gpiod_line *columnclock;
struct gpiod_line *latch;
struct gpiod_line *clrbar;
struct gpiod_line *fe;
struct gpiod_line *rowdata;
struct gpiod_line *rowclock;

// size of the display: we have 7 rows and 9 columns of 8bit size

unsigned char characters[128 * 8];

void delayMicroseconds(int delay){
  usleep(delay);
}

void delay(int delay){
  usleep(1000*delay);
}

void abort(){
  gpiod_line_release(data);
  gpiod_line_release(latch);
  gpiod_line_release(columnclock);
  gpiod_line_release(clrbar);
  gpiod_line_release(fe);
  gpiod_line_release(rowdata);
  gpiod_line_release(rowclock);
  gpiod_chip_close(chip);
  exit(0);
}

struct gpiod_line * setAsOutput(int pin){
  struct gpiod_line *gpio_line;
  int result;

  gpio_line = gpiod_chip_get_line(chip, pin);
  if (!gpio_line) {
    gpiod_chip_close(chip);
    perror("Get line failed\n");
    exit(1);
  }
  result = gpiod_line_request_output(gpio_line, CONSUMER, 0);
  if (result < 0) {
    perror("Request line as output failed\n");
    abort();
  }
  return(gpio_line);
}

void init_hardware()
{
  char *chipname = "gpiochip0";
  /* open the chip */
  chip = gpiod_chip_open_by_name(chipname);
  if (!chip) {
    perror("Open chip failed\n");
    exit(1);
  }
  data = setAsOutput(dataPin);
  latch = setAsOutput(latchPin);
  columnclock = setAsOutput(columnclockPin);
  clrbar = setAsOutput(clrbarPin);
  fe = setAsOutput(fePin);
  rowdata = setAsOutput(rowdataPin);
  rowclock = setAsOutput(rowclockPin);
}

void writeState(struct gpiod_line *line, unsigned int state){
  int result;
  result = gpiod_line_set_value(line, state);
  if (result < 0){
    perror("Set line output failed\n");
    abort();
  }
}

void out_byte(struct gpiod_line * dLine, struct gpiod_line * cLine, int val){
  int i;
  for (i = 0; i < 8; i++) {
    writeState(cLine, LOW);
    writeState(dLine, ((0x80 & (val << i)) == 0x80) ? HIGH : LOW);
    delayMicroseconds(10);
    writeState(cLine, HIGH);
    delayMicroseconds(10);
  }
}

void out_byte_lsb(struct gpiod_line * dLine, struct gpiod_line * cLine, int val){
  int i;
  for (i = 0; i < 8; i++) {
    writeState(cLine, LOW);
    writeState(dLine, ((0x1 & (val >> i)) == 0x1) ? HIGH : LOW);
    delayMicroseconds(10);
    writeState(cLine, HIGH);
    delayMicroseconds(10);
  }
}

void out_byte_watch(struct gpiod_line * dLine, struct gpiod_line * cLine, struct gpiod_line * lLine, int val){
  int i;
  for (i = 0; i < 8; i++) {
    writeState(cLine, LOW);
    writeState(dLine, ((0x80 & (val << i)) == 0x80) ? HIGH : LOW);
    delayMicroseconds(10);
    writeState(cLine, HIGH);
    delayMicroseconds(10);
    writeState(lLine, HIGH);
    delayMicroseconds(10);
    writeState(lLine, LOW);
    delay(500);
  }
}

void switch_off(){
  // disable demultiplexer
  writeState(fe, LOW);
  // reset shift registers
  writeState(clrbar, LOW);
  delayMicroseconds(10);
  writeState(clrbar, HIGH);
  delayMicroseconds(10);
}

void * thread_show_buffer_on_led(void *buffer){
  int             nx = ROWS;
  int             ny = COLUMNS;
  int             i, j;
  unsigned char   x;
  int             block, inside, pos;
  unsigned char  *buff = (unsigned char *) buffer;
  writeState(latch, LOW);
  while (1) {
    for (j = 0; j < nx; j++) {
      for (i = 0; i < ny; i++) {
	block = i / 3;
	inside = i - 3 * block;
	pos = 3 * block + 2 - inside;
	x = *(buff + ROWS * pos + j);
	out_byte_lsb(data, columnclock, x ^ 0xff);
	if (inside == 2) {
	  x = j + 8 * block;
	  out_byte(rowdata, rowclock, x);
	  writeState(latch, HIGH);
	  delayMicroseconds(50);
	  writeState(latch, LOW);
	}
      }
    }
  }
}

int main(int argc, char *argv[]) {
  unsigned char   characterset[CHARSET_SIZE + 1];
  char            textbuffer[LCOLUMNS];

  // initialise the hardware
  init_hardware();

  // initialise text
  // read_character_set(&characterset[0]);
  memcpy(characterset, font, 2048);

  // create the display thread
  create_display_thread(thread_show_buffer_on_led);

  // write some text into the large buffer
  strcpy(textbuffer, "        Hello, World !!!       ");
  write_string(strlen(textbuffer), (unsigned char *) large_display_buffer, characterset, textbuffer);

  // make a sliding window
  int             copy = COLUMNS * ROWS;	// fill the full display buffer

  for (int replay = 0; replay < 3; replay++) {
    // initialize the text
    write_string(strlen(textbuffer), (unsigned char *) large_display_buffer, characterset, textbuffer);
    memcpy((unsigned char *) display_buffer, ((unsigned char *) large_display_buffer), copy);
    for (int i = 0; i < strlen(textbuffer); i++) {
      for (int ii = 0; ii < 8; ii++) {
	for (int jj = 0; jj < ROWS; jj++) {
	  shiftleftonce(jj, ROWS, LCOLUMNS, ((unsigned char *)large_display_buffer));
	}
	memcpy((unsigned char *) display_buffer, ((unsigned char *) large_display_buffer), copy);
	delayMicroseconds(50000);
      }
    }
  }
  switch_off();
  abort();
}
